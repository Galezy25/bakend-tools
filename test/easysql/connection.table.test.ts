import Column from '../../src/easysql/column';
import Connection from '../../src/easysql/connection';
import {
    Constraint,
    ForeignKey,
    PrimaryKey,
} from '../../src/easysql/constraint';

const mockQuery = jest.fn(
    (__query, callback: (err: any, results: any, fields: any) => void) => {
        callback(undefined, [], []);
    }
);
jest.mock('mysql', () => {
    return {
        createPool: jest.fn(_config => ({
            query: mockQuery,
            end: jest.fn(),
        })),
        escape: jest.fn((value: any) => {
            return typeof value === 'string' ? `'${value}'` : value + '';
        }),
    };
});

describe('Connection tests', () => {
    let connection = new Connection({
        user: 'tester',
        database: 'test',
    });

    const tableTestsName = 'table_tests';
    const table = connection.table(tableTestsName);

    beforeEach(() => {
        mockQuery.mockClear();
    });

    test('Secure query', async () => {
        const queryTemplate = 'this is a test named {{ name }} {{ num }}';
        expect(
            connection.secureQuery(queryTemplate, { name: 'Simple', num: 25 })
        ).toBe("this is a test named 'Simple' 25");
        expect(
            connection.secureQuery(queryTemplate, {
                name: 'CURRENT_TIMESTAMP',
                num: null,
            })
        ).toBe('this is a test named CURRENT_TIMESTAMP NULL');
        expect(
            connection.secureQuery(queryTemplate, { name: true, num: NaN })
        ).toBe('this is a test named true NaN');
        expect(
            connection.secureQuery(queryTemplate, {
                name: 'SELECT * FROM users',
                num: undefined,
            })
        ).toBe("this is a test named 'SELECT * FROM users' DEFAULT");
    });

    test('Create table', async () => {
        const columns: Column[] = [
            new Column('id', {
                dataType: 'INT',
                autoIncremet: true,
            }),
            new Column('name', {
                dataType: ['VARCHAR', 150],
            }),
            new Column('email', {
                dataType: ['VARCHAR', 30],
                null: false,
                unique: true,
            }),
            new Column('leader', {
                dataType: 'INT',
                null: true,
                default: 'NULL',
            }),
            new Column('active', {
                dataType: 'BOOLEAN',
                default: true,
            }),
        ];
        const constraints: Constraint[] = [
            new PrimaryKey('id'),
            new ForeignKey({
                colName: 'leader',
                tableRef: tableTestsName,
                columnRef: 'id',
            }),
        ];
        let table = await connection.createTable(
            tableTestsName,
            ...columns,
            ...constraints
        );
        expect(table.name).toBe(tableTestsName);
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('Find, Find One of a empty table', async () => {
        await table.find({});
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
        await expect(table.findOne({ id: 1 })).rejects.toMatchObject({
            code: 404,
        });
        expect(mockQuery.mock.calls[1][0]).toMatchSnapshot();
    });

    test('Add, modify and drop column', async () => {
        await table.addColumn(
            new Column('phone', {
                dataType: 'BIGINT',
            })
        );
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
        await table.modifyColumn(
            new Column('phone', {
                dataType: ['VARCHAR', 17],
            })
        );
        expect(mockQuery.mock.calls[1][0]).toMatchSnapshot();
        await table.dropColumn('phone');
        expect(mockQuery.mock.calls[2][0]).toMatchSnapshot();
    });

    test('Create rows', async () => {
        await table.create({
            name: 'test_1',
            email: 'test_1@test.test',
        });
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();

        let toRegister: any[] = [];
        for (let index = 2; index <= 200; index++) {
            toRegister.push({
                name: 'test_' + index,
                email: 'test_' + index + '@test.test',
                leader: index > 10 ? index % 10 || 1 : null,
                otherValue: index,
            });
        }
        await table.create(toRegister, ['name', 'email', 'leader']);
        expect(mockQuery.mock.calls[1][0]).toMatchSnapshot();
    });

    test('Find rows', async () => {
        await table.find(
            {
                _sort: 'id:DESC',
                _limit: 25,
                _start: 25,
                leader_gteq: 8,
                [tableTestsName + '_inner']: tableTestsName + '.id:this.leader',
            },
            ['this.*', tableTestsName + '.name AS leader_name']
        );
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();

        await table.find({
            email_like: 'test_2%',
            leader_in: '2,5',
        });
        expect(mockQuery.mock.calls[1][0]).toMatchSnapshot();
    });

    test('Update row', async () => {
        await table.update({ id: 100 }, { name: 'test_100_2' });
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('Delete row', async () => {
        await table.delete({ id: 100 });
        expect(mockQuery.mock.calls[0][0]).toMatchSnapshot();
    });
});
