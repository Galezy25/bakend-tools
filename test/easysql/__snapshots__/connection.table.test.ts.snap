// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection tests Add, modify and drop column 1`] = `"ALTER TABLE table_tests ADD COLUMN phone BIGINT"`;

exports[`Connection tests Add, modify and drop column 2`] = `"ALTER TABLE table_tests MODIFY COLUMN phone VARCHAR(17)"`;

exports[`Connection tests Add, modify and drop column 3`] = `"ALTER TABLE table_tests DROP COLUMN phone"`;

exports[`Connection tests Create rows 1`] = `" INSERT INTO table_tests (name,email) VALUES ('test_1', 'test_1@test.test');"`;

exports[`Connection tests Create rows 2`] = `" INSERT INTO table_tests (name,email,leader) VALUES ('test_2', 'test_2@test.test', NULL),('test_3', 'test_3@test.test', NULL),('test_4', 'test_4@test.test', NULL),('test_5', 'test_5@test.test', NULL),('test_6', 'test_6@test.test', NULL),('test_7', 'test_7@test.test', NULL),('test_8', 'test_8@test.test', NULL),('test_9', 'test_9@test.test', NULL),('test_10', 'test_10@test.test', NULL),('test_11', 'test_11@test.test', 1),('test_12', 'test_12@test.test', 2),('test_13', 'test_13@test.test', 3),('test_14', 'test_14@test.test', 4),('test_15', 'test_15@test.test', 5),('test_16', 'test_16@test.test', 6),('test_17', 'test_17@test.test', 7),('test_18', 'test_18@test.test', 8),('test_19', 'test_19@test.test', 9),('test_20', 'test_20@test.test', 1),('test_21', 'test_21@test.test', 1),('test_22', 'test_22@test.test', 2),('test_23', 'test_23@test.test', 3),('test_24', 'test_24@test.test', 4),('test_25', 'test_25@test.test', 5),('test_26', 'test_26@test.test', 6),('test_27', 'test_27@test.test', 7),('test_28', 'test_28@test.test', 8),('test_29', 'test_29@test.test', 9),('test_30', 'test_30@test.test', 1),('test_31', 'test_31@test.test', 1),('test_32', 'test_32@test.test', 2),('test_33', 'test_33@test.test', 3),('test_34', 'test_34@test.test', 4),('test_35', 'test_35@test.test', 5),('test_36', 'test_36@test.test', 6),('test_37', 'test_37@test.test', 7),('test_38', 'test_38@test.test', 8),('test_39', 'test_39@test.test', 9),('test_40', 'test_40@test.test', 1),('test_41', 'test_41@test.test', 1),('test_42', 'test_42@test.test', 2),('test_43', 'test_43@test.test', 3),('test_44', 'test_44@test.test', 4),('test_45', 'test_45@test.test', 5),('test_46', 'test_46@test.test', 6),('test_47', 'test_47@test.test', 7),('test_48', 'test_48@test.test', 8),('test_49', 'test_49@test.test', 9),('test_50', 'test_50@test.test', 1),('test_51', 'test_51@test.test', 1),('test_52', 'test_52@test.test', 2),('test_53', 'test_53@test.test', 3),('test_54', 'test_54@test.test', 4),('test_55', 'test_55@test.test', 5),('test_56', 'test_56@test.test', 6),('test_57', 'test_57@test.test', 7),('test_58', 'test_58@test.test', 8),('test_59', 'test_59@test.test', 9),('test_60', 'test_60@test.test', 1),('test_61', 'test_61@test.test', 1),('test_62', 'test_62@test.test', 2),('test_63', 'test_63@test.test', 3),('test_64', 'test_64@test.test', 4),('test_65', 'test_65@test.test', 5),('test_66', 'test_66@test.test', 6),('test_67', 'test_67@test.test', 7),('test_68', 'test_68@test.test', 8),('test_69', 'test_69@test.test', 9),('test_70', 'test_70@test.test', 1),('test_71', 'test_71@test.test', 1),('test_72', 'test_72@test.test', 2),('test_73', 'test_73@test.test', 3),('test_74', 'test_74@test.test', 4),('test_75', 'test_75@test.test', 5),('test_76', 'test_76@test.test', 6),('test_77', 'test_77@test.test', 7),('test_78', 'test_78@test.test', 8),('test_79', 'test_79@test.test', 9),('test_80', 'test_80@test.test', 1),('test_81', 'test_81@test.test', 1),('test_82', 'test_82@test.test', 2),('test_83', 'test_83@test.test', 3),('test_84', 'test_84@test.test', 4),('test_85', 'test_85@test.test', 5),('test_86', 'test_86@test.test', 6),('test_87', 'test_87@test.test', 7),('test_88', 'test_88@test.test', 8),('test_89', 'test_89@test.test', 9),('test_90', 'test_90@test.test', 1),('test_91', 'test_91@test.test', 1),('test_92', 'test_92@test.test', 2),('test_93', 'test_93@test.test', 3),('test_94', 'test_94@test.test', 4),('test_95', 'test_95@test.test', 5),('test_96', 'test_96@test.test', 6),('test_97', 'test_97@test.test', 7),('test_98', 'test_98@test.test', 8),('test_99', 'test_99@test.test', 9),('test_100', 'test_100@test.test', 1),('test_101', 'test_101@test.test', 1),('test_102', 'test_102@test.test', 2),('test_103', 'test_103@test.test', 3),('test_104', 'test_104@test.test', 4),('test_105', 'test_105@test.test', 5),('test_106', 'test_106@test.test', 6),('test_107', 'test_107@test.test', 7),('test_108', 'test_108@test.test', 8),('test_109', 'test_109@test.test', 9),('test_110', 'test_110@test.test', 1),('test_111', 'test_111@test.test', 1),('test_112', 'test_112@test.test', 2),('test_113', 'test_113@test.test', 3),('test_114', 'test_114@test.test', 4),('test_115', 'test_115@test.test', 5),('test_116', 'test_116@test.test', 6),('test_117', 'test_117@test.test', 7),('test_118', 'test_118@test.test', 8),('test_119', 'test_119@test.test', 9),('test_120', 'test_120@test.test', 1),('test_121', 'test_121@test.test', 1),('test_122', 'test_122@test.test', 2),('test_123', 'test_123@test.test', 3),('test_124', 'test_124@test.test', 4),('test_125', 'test_125@test.test', 5),('test_126', 'test_126@test.test', 6),('test_127', 'test_127@test.test', 7),('test_128', 'test_128@test.test', 8),('test_129', 'test_129@test.test', 9),('test_130', 'test_130@test.test', 1),('test_131', 'test_131@test.test', 1),('test_132', 'test_132@test.test', 2),('test_133', 'test_133@test.test', 3),('test_134', 'test_134@test.test', 4),('test_135', 'test_135@test.test', 5),('test_136', 'test_136@test.test', 6),('test_137', 'test_137@test.test', 7),('test_138', 'test_138@test.test', 8),('test_139', 'test_139@test.test', 9),('test_140', 'test_140@test.test', 1),('test_141', 'test_141@test.test', 1),('test_142', 'test_142@test.test', 2),('test_143', 'test_143@test.test', 3),('test_144', 'test_144@test.test', 4),('test_145', 'test_145@test.test', 5),('test_146', 'test_146@test.test', 6),('test_147', 'test_147@test.test', 7),('test_148', 'test_148@test.test', 8),('test_149', 'test_149@test.test', 9),('test_150', 'test_150@test.test', 1),('test_151', 'test_151@test.test', 1),('test_152', 'test_152@test.test', 2),('test_153', 'test_153@test.test', 3),('test_154', 'test_154@test.test', 4),('test_155', 'test_155@test.test', 5),('test_156', 'test_156@test.test', 6),('test_157', 'test_157@test.test', 7),('test_158', 'test_158@test.test', 8),('test_159', 'test_159@test.test', 9),('test_160', 'test_160@test.test', 1),('test_161', 'test_161@test.test', 1),('test_162', 'test_162@test.test', 2),('test_163', 'test_163@test.test', 3),('test_164', 'test_164@test.test', 4),('test_165', 'test_165@test.test', 5),('test_166', 'test_166@test.test', 6),('test_167', 'test_167@test.test', 7),('test_168', 'test_168@test.test', 8),('test_169', 'test_169@test.test', 9),('test_170', 'test_170@test.test', 1),('test_171', 'test_171@test.test', 1),('test_172', 'test_172@test.test', 2),('test_173', 'test_173@test.test', 3),('test_174', 'test_174@test.test', 4),('test_175', 'test_175@test.test', 5),('test_176', 'test_176@test.test', 6),('test_177', 'test_177@test.test', 7),('test_178', 'test_178@test.test', 8),('test_179', 'test_179@test.test', 9),('test_180', 'test_180@test.test', 1),('test_181', 'test_181@test.test', 1),('test_182', 'test_182@test.test', 2),('test_183', 'test_183@test.test', 3),('test_184', 'test_184@test.test', 4),('test_185', 'test_185@test.test', 5),('test_186', 'test_186@test.test', 6),('test_187', 'test_187@test.test', 7),('test_188', 'test_188@test.test', 8),('test_189', 'test_189@test.test', 9),('test_190', 'test_190@test.test', 1),('test_191', 'test_191@test.test', 1),('test_192', 'test_192@test.test', 2),('test_193', 'test_193@test.test', 3),('test_194', 'test_194@test.test', 4),('test_195', 'test_195@test.test', 5),('test_196', 'test_196@test.test', 6),('test_197', 'test_197@test.test', 7),('test_198', 'test_198@test.test', 8),('test_199', 'test_199@test.test', 9),('test_200', 'test_200@test.test', 1);"`;

exports[`Connection tests Create table 1`] = `
"CREATE TABLE table_tests(
        id INT AUTO_INCREMENT,
name VARCHAR(150),
email VARCHAR(30) NOT NULL UNIQUE,
leader INT NULL DEFAULT NULL,
active BOOLEAN DEFAULT true,
CONSTRAINT PRIMARY KEY (id),
CONSTRAINT FOREIGN KEY (leader) REFERENCES table_tests(id));"
`;

exports[`Connection tests Delete row 1`] = `" DELETE FROM table_tests  WHERE id = 100 "`;

exports[`Connection tests Find rows 1`] = `" SELECT this.*, table_tests.name AS leader_name FROM table_tests this  INNER JOIN table_tests ON table_tests.id=this.leader WHERE leader >= 8  ORDER BY id DESC LIMIT 25, 25 ; "`;

exports[`Connection tests Find rows 2`] = `" SELECT * FROM table_tests this  WHERE email LIKE 'test_2%' AND leader IN ('2','5') ; "`;

exports[`Connection tests Find with or/and 1`] = `" SELECT * FROM table_tests this  WHERE (email LIKE '%@test.net' OR email LIKE '%@subdomain.test.net' OR email LIKE '%@subdomain2.test.net') AND leader IN ('2','5') ; "`;

exports[`Connection tests Find, Find One of a empty table 1`] = `" SELECT * FROM table_tests this ; "`;

exports[`Connection tests Find, Find One of a empty table 2`] = `" SELECT *  FROM table_tests this  WHERE id = 1 "`;

exports[`Connection tests Update row 1`] = `" UPDATE table_tests SET name='test_100_2'  WHERE id = 100 "`;
